peps/pep-0728.rst

```diff
@@ -0,0 +1,633 @@
+        [key: string]: string | number
+    }
+
+    interface B extends A {
+        foo: number
+    }
+
+    const x: B = {value: "asd", foo: 12}
+    function mut(v: A) {
+        v.foo = "asd"
+    }
+    mut(x)
+    console.log(x)  // {"value": "asd", "foo": "asd"}
+
+Because ``A`` is a subtype of ``B``, ``mut`` accepts ``x`` and mutates the value
+of the property ``foo``, changing it to a ``string``. However, ``x`` is defined
+to be ``B``, which expects ``foo`` to be a ``number``. As a result, the value of
+``x`` in runtime is no longer consistent with its type during static analysis.
+
+To support this in Python, we need to determine if we need this level of
+flexibility to open up this hole of soundness.
+
+This PEP as written is more restrictive to match the type consistency rules
+defined in :pep:`589`. A more relaxed design may, for example, no longer
+prohibit subclasses of a TypedDict type that defines ``__extra__`` from adding
+more required fields::
+
+    class IntDict(TypedDict):
+        __extra__: int
+
+    class IntDictWithNum(IntDict):
+        num: int
+
+    def f(x: IntDict) -> None:
+        del x["num"]  # OK
+    
+    required_num: IntDictWithNum = {"num": 1}
+    f(required_num)  # OK
+    required_num["num"]  # KeyError in runtime
+
+Should ``__extra__`` Exclude the Known Keys?
+--------------------------------------------
+
+``__extra__`` specifies the value type for fields of keys that are *unknown* to
+the TypedDict type. So the value type of any known field is not necessarily
+consistent with ``__extra__``'s type, and ``__extra__``'s type is not
+necessarily consistent with the value types of all known fields.
+
+This differs from TypeScript's Index Signatures [8]_ syntax, which requires all
+properties' types to match the string index's type. For example:
+
+.. code-block:: typescript
+
+    interface MovieWithExtraNumber {
+        name: string // Property 'name' of type 'string' is not assignable to 'string' index type 'number'.
+        [index: string]: number
+    }
+
+    interface MovieWithExtraNumberOrString {
+        name: string // OK
+        [index: string]: number | string
+    }
+
+This is a known limitation is discussed in TypeScript's issue tracker [9]_,
+where it is suggested that there should be a way to exclude the defined keys
+from the index signature, so that it is possible to define a type like
+``MovieWithExtraNumber``.
+
+A workaround is to use intersection:
+
+.. code-block:: typescript
+
+    interface MovieExtras {
+        [index: string]: number
+    }
+
+    interface MovieBase {
+        name: string
+    }
+
+    type Movie = MovieExtras & MovieBase
+
+However, this does not work for assignment:
+
+.. code-block:: typescript
+
+    const movie: Movie = {
+        name: "No Country for Old Men", // Type 'string' is not assignable to type 'number'
+        year: 2007,
+    }
+
+How Else Will This Interact with PEP 692?
+-----------------------------------------
```

Comment:
It's not clear what the concrete open issue is here.

Reply:
Got it removed. I think the "Interaction with `Unpack`" section should be sufficient.